presto-password-authenticators/src/main/java/io/prestosql/plugin/password/filebasedauthenticator/User.java

```diff
@@ -0,0 +1,186 @@
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkCryptPassword;
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkMD5Password;
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkPlainTextPassword;
+import static io.prestosql.plugin.password.filebasedauthenticator.EncryptionHelper.checkSHAPassword;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class User
+{
+    private final String name;
+    private final String password;
+
+    // empty encryptionAlgo implies that correct algo is yet to be set
+    // Setting the algo to future use is an optimization to prevent going through all supported algos for every authentication call
+    private Optional<EncryptionAlgo> encryptionAlgo = Optional.empty();
+
+    public static final String BASIC_AUTHENTICATION_PREFIX = "Basic ";
+
+    public User(
+            String name,
+            String password)
+    {
+        this(name, password, true);
+    }
+
+    public User(String name, String password, boolean storedUser)
+    {
+        this.name = requireNonNull(name, "User name is empty");
+        requireNonNull(password, "User name is empty");
+
+        // For BCrypt htpasswd generates password with $2y prefix while jbcrypt requires $2a
+        if (storedUser && password.startsWith("$2y")) {
+            this.password = password.replace("$2y", "$2a");
+        }
+        else {
+            this.password = password;
+        }
+
+        if (!storedUser) {
+            // Store UNKNOWN in algo for the user under check, only stored passwords can have encryption
+            encryptionAlgo = Optional.of(UNKNOWN);
+        }
+    }
+
+    public String getName()
+    {
+        return name;
+    }
+
+    public String getPassword()
+    {
+        return password;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        User that = (User) o;
+
+        if (!Objects.equals(this.name, that.name)) {
+            return false;
+        }
+
+        // The user object created for the user submitted to server has encryptionAlgo set as UNKNOWN
+        if (!this.encryptionAlgo.isPresent() || this.encryptionAlgo.get() != UNKNOWN) {
+            return authenticate(that.password, this);
+        }
+        else {
+            return authenticate(this.password, that);
+        }
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(name);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("user", name)
+                .add("password", password)
+                .toString();
+    }
+
+    public static User getUser(String httpAuthHeader)
```

Comment:
This method is not used. Same for `decodeCredentials` and `BASIC_AUTHENTICATION_PREFIX`.